pkg/arvo/app/graph-store.hoon

```diff
@@ -0,0 +1,544 @@
+    ==
+    ::
+    ++  add-graph
+      |=  [=resource:store =graph:store mark=(unit mark:store)]
+      ^-  (quip card _state)
+      ?<  (~(has by archive) resource)
+      ?<  (~(has by graphs) resource)
+      ?>  (validate-graph graph mark)
+      :_  %_  state
+              graphs       (~(put by graphs) resource [graph mark])
+              update-logs  (~(put by update-logs) resource (gas:orm-log ~ ~))
+          ==
+      %-  zing
+      :~  (give [/updates /keys ~] [%add-graph resource graph mark])
+          ?~  mark  ~
+          =/  wire  (weld /graph (en-path:res resource))
+          =/  =rave:clay  [%sing %b [%da now.bowl] /[u.mark]]
+          [%pass wire %arvo %c %warp our.bowl [%home `rave]]~
+      ==
+    ::
+    ++  remove-graph
+      |=  =resource:store
+      ^-  (quip card _state)
+      ?<  (~(has by archive) resource)
+      ?>  (~(has by graphs) resource)
+      :-  (give [/updates /keys ~] [%remove-graph resource])
+      %_  state
+          graphs       (~(del by graphs) resource)
+          update-logs  (~(del by update-logs) resource)
+      ==
+    ::
+    ++  add-nodes
+      |=  $:  =time
+              =resource:store
+              nodes=(map index:store node:store)
+          ==
+      ^-  (quip card _state)
+      |^
+      =/  [=graph:store mark=(unit mark:store)]
+        (~(got by graphs) resource)
+      =/  =update-log:store  (~(got by update-logs) resource)
+      =.  update-log
+        (put:orm-log update-log time [%0 time [%add-nodes resource nodes]])
+      ::
+      :-  (give [/updates]~ [%add-nodes resource nodes])
+      %_  state
+          update-logs  (~(put by update-logs) resource update-log)
+          graphs
+        %+  ~(put by graphs)
+          resource
+        [(add-node-list resource graph mark ~(tap by nodes)) mark]
+      ==
+      ::
+      ++  add-node-list
+        |=  $:  =resource:store
+                =graph:store
+                mark=(unit mark:store)
+                node-list=(list [index:store node:store])
+            ==
+        ^-  graph:store
+        ?~  node-list  graph
+        =*  index  -.i.node-list
+        =*  node   +.i.node-list
+        %_  $
+            node-list  t.node-list
+            graph      (add-node-at-index graph index node ~ mark)
+        ==
+      ::
+      ++  add-node-at-index
+        |=  $:  =graph:store
+                =index:store
+                =node:store
+                parent-hash=(unit hash:store)
+                mark=(unit mark:store)
+            ==
+        ^-  graph:store
+        ?~  index  graph
+        ?>  (validate-graph (gas:orm ~ [i.index node]~) mark)
+        =*  atom   i.index
+        %^  put:orm
+            graph
+          atom
+        ::  add child
+        ::
+        ?~  t.index
+          =*  p  post.node
+          =/  =validated-portion:store
+            [parent-hash author.p index.p time-sent.p contents.p]
+          =/  =hash:store  (mug validated-portion)
```

Comment:
Why? @philipcmonk Thoughts?

Reply:
I'm not sure exactly what this hash is used for, but if it's for any kind of validation beyond a simple error-correcting code in case of transmission error, `mug` is probably insufficient.  `mug` is 31 bits and hence has a trivial preimage attack:

```
|-(?:(=(123 (mug eny)) eny $(eny +(eny))))

0vla.s4dl5.ha9lj.nomc1.csuv8.fkmj7.5jkmo.q0d3j.7uihn.2imun.99lso.ktdlm.ivtjs.q23cr.95lsi.ladgs.8n8aq.lf3pd.b0l3m.5immq.3u9sv
```

This took a couple minutes to generate something which mugs to "123".

Incidentally, mug is also bad as a key in a map if you don't account for collisions, since by the birthday problem, you only need about 70k entries in the map before you're likely to have a collision.

Reply:
@philipcmonk Do you have a suggestion for which hash to use? Your explanation makes sense. I'm thinking `sham` would be adequate, with its 128-bit noun hash.

Reply:
(not getting into other recent issues discovered in mug)

Reply:
`sham` is good